REmove AC specific flag and reunify with HOPs

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #169134
* __->__ #169132
* #169131


Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

tugsbayasgalan committed:
REmove AC specific flag and reunify with HOPs

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -3298,12 +3324,32 @@ def _call_function(
+                tx, fn_vt, fn_args_vt, kwargs, description, subgraph_name=subgraph_name
+            )
+        finally:
+            tx.output.current_tracer.under_activation_checkpoint = orig_val
```
I think we can get rid of `under_activation_checkpoint` entirely from the codebase. So need to overwrite this method.

anijain2305 reviewed (commented):
Minor comments, but these are important to address before merging the PR so that others can collaborate on top of these.

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 reviewed (commented):
Requesting minor changes (which hopefully are minor, and do not blow up the CI).

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

anijain2305 reviewed (approved):
Ok this is cleaner, can you add just address the two text comments?

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "REmove AC specific flag and reunify with HOPs"



Replaces checkpoint-specific side effects mechanism with a unified `allow_side_effects_with_extra_outputs` flag that works for all HOPs. We do this by reusing `restore_side_effects` flag as much as possible.  Based on my understanding, this flag works by:

**When `restore_side_effects = True` (default)**:
- Side effects data structure is saved before tracing and restored after
- Intermediate tensors/symints are NOT captured as extra outputs
- Used by most HOPs

**When `restore_side_effects = False`**:
- Side effects data structure is NOT restored
- Intermediate tensors/symints ARE captured and returned as extra outputs
- Used by `invoke_subgraph` (always) and `checkpoint` (when `config.skip_fwd_side_effects_in_bwd_under_checkpoint=True`)

We had to keep some `under_activation_checkpointing` due to legacy FSDP changes. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]
